doc/remediation-templating.md

```diff
@@ -0,0 +1,342 @@
+
+#UseDNS no
+#PidFile /var/run/sshd.pid
+#MaxStartups 10:30:100
+#PermitTunnel no
+...
+```
+Since the url encode doesn't encode ASCII alpha-numerical nor underscore characters, and
+we can later substitute the place holder after the url encoding from `VAR_SSHD_IDLE_TIMEOUT_VALUE`
+to `{{sshd_idle_timeout_value | findValue | urlquery}}`, so that it can be post-process them with the appropriate
+escaped instances in the compliance operator. The following code is the example of final result:
+```xml
+<fix rule="sshd_set_idle_timeout" complexity="low" disruption="low" reboot="false" strategy="restrict">---
+apiVersion: machineconfiguration.openshift.io/v1
+kind: MachineConfig
+spec:
+  config:
+    ignition:
+      version: 3.1.0
+    storage:
+      files:
+      - contents:
+          source: data:, `PrintLastLog%20yes%0A%23TCPKeepAlive%20yes%0APermitUserEnvironment%20no%0ACompression
+          %20no%0AClientAliveInterval%20{{"sshd_idle_timeout_value" | findValue | urlquery}}%0AClientAliveCountMax%200%0A%0A%23UseDNS%2
+          0no%0A%23PidFile%20%2Fvar%2Frun%2Fsshd.pid%0A%23MaxStartups%2010%3A30%3A100%0A%23PermitTunnel%20no`
+        mode: 0600
+        path: /etc/ssh/sshd_config
+        overwrite: true
+</fix>
+```
+
+There will be some change in the aggregator since we brought a marker XCCDF variable into
+the remediation content, we need to substitute all the placeholder with corresponding values
+before we create the `ComplianceRemediation` object. we need to fetch all the variables that 
+are in the scan result `ConfigMap` for all the nodes with consistency check(Note that all the 
+tailored set values are parsed to the resulting CM), and then since we have XCCDF values marked 
+using `{{}}` in the remediation rules there is a Go Library called `Template` that we can 
+use to match, url-encode, and substitute the value with the XCCDF variable set in the Tailored Profile 
+from the result `ConfigMap` and then to create the `ComplianceRemediation` object with 
+corresponding values.
+
+Proof of concept using golang templating https://golang.org/pkg/text/template/#hdr-Functions 
+```go
+var value_dic = map[string]string{
+	"the_value_you_want_to_define":  "3600",
+	"the_value_you_want_to_define2": "1111",
+	"the_value_you_want_to_define3": "2222",
+	"the_value_you_want_to_define4": "3333",
+}
+
+func findValue(format string) string {
+	value, ok := value_dic[format]
+	if ok {
+		return value
+	} else {
+		return "NoValue"
+	}
+
+}
+
+func main() {
+	MachineConfigCM := "data:,...%0APrintMotd%20no%0A%0APrintLastLog%20yes%0A%23TCPKeepAlive%20yes%0APermitUserEnvironment%20no%0ACompression%20no%0AClientAliveInterval%20{{\"the_value_you_want_to_define65\" | findValue | urlquery}}%0AClientAliveCountMax%200%0A%23UseDNS%20no%0A%23PidFile%20/var/run/sshd.pid%0A%23MaxStartups%2010%3A30%3A100%0A%23PermitTunnel%20no%0A%23ChrootDirectory%20none%0A%23..."
+	t, err := template.New("text").Funcs(template.FuncMap{"findValue": findValue}).
```

Comment:
I like the usage of FuncMap here! I hadn't thought about using this function but it makes a lot of sense!

Reply:
I'd say this should have a more specific name, such as `findXCCDFValue` or something of the sort. @Vincent056 @jhrozek @mrogers950 thoughts on this?

Reply:
Yes, I agree with that.

Reply:
I am going to change to that on the design doc

Reply:
could you change it? seems that it's still `findValue` as opposed to a more specific name